backend/uclapi/libcal/views.py

```diff
@@ -0,0 +1,407 @@
+    params["email"] = user.mail if user.mail else user.email
+    return _libcal_request_forwarder(
+        "/1.1/space/bookings",
+        request,
+        LibCalPersonalBookingsGETSerializer(data=params),
+        'bookings',
+        **kwargs
+    )
+
+
+@api_view(["POST"])
+@parser_classes([JSONParser])
+@uclapi_protected_endpoint(personal_data=True, required_scopes=['libcal_write'], last_modified_redis_key=None)
+def reserve(request, *args, **kwargs):
+    """Reserves a space/seat on behalf of a user. Requires OAuth permissions"""
+    user = kwargs['token'].user
+    # NOTE: serializer will catch if this is empty (default is '').
+    request.data["email"] = user.mail if user.mail else user.email
+    request.data["fname"] = user.given_name
+    request.data["lname"] = user.sn
+    return _libcal_request_forwarder(
+        "/1.1/space/reserve",
+        request,
+        LibCalReservationPOSTSerializer(data=request.data),
+        'bookings',
+        'POST',
+        **kwargs
+    )
+
+
+@api_view(["POST"])
+@uclapi_protected_endpoint(personal_data=True, required_scopes=['libcal_write'], last_modified_redis_key=None)
+def cancel(request, *args, **kwargs):
+    """Cancels a booking on behalf of a user."""
+    # The /1.1/space/cancel endpoint simply accepts a booking ID(s) as a way of cancelling.
+    # This format of the ID seems to be r"^cs_\w{8}$" thus allowing an attacker to cancel arbitrary bookings by process
+    # of enumeration.
+    # We add an additional level of security by checking the personal OAuth token, to check that the app cancelling a
+    # booking has permission to delete the booking associated with the user.
+    user = kwargs['token'].user
+    email = user.mail if user.mail else user.email
+    if not email:
+        uclapi_response = JsonResponse({
+            "ok": False,
+            "error": "This booking cannot be cancelled as this user has no valid email address."
+        }, custom_header_data=kwargs)
+        uclapi_response.status_code = 500
+        return uclapi_response
+
+    booking_response = _libcal_request_forwarder(
+        "/1.1/space/booking",
+        request,
+        LibCalBookingIdListSerializer(data=request.query_params),
+        'data',
+        **kwargs
+    )
+    data = json.loads(booking_response.content.decode('utf-8'))
+    if "data" not in data:
+        # An error occured with this request, forward to user.
+        return booking_response
+    else:
+        # Only send the booking IDs to LibCal that correspond to the user
+        legit_bids = ''
+        for booking in data["data"]:
+            if booking["email"] == email:
+                legit_bids += booking["bookId"] + ','
+        if legit_bids:
+            legit_bids = legit_bids[:-1]  # Remove last ','
+        else:
+            uclapi_response = JsonResponse({
+                "ok": False,
+                "error": "No bookings were found to be deleted"
```

Comment:
Rephrase this? Maybe: "No bookings matched the IDs provided and so no bookings were deleted"

Reply:
Done